aidial_adapter_openai/exception_handlers.py

```diff
@@ -1,56 +1,81 @@
-        if "Content-Encoding" in headers:
-            del headers["Content-Encoding"]
+def to_adapter_exception(exc: Exception) -> AdapterException:
+    if isinstance(exc, APIStatusError):
+        # Non-streaming errors reported by `openai` library via this exception
+        r = exc.response
+        httpx_headers = r.headers
 
-        return Response(
-            content=r.content,
+        # httpx library (used by openai) automatically sets
+        # "Accept-Encoding:gzip,deflate" header in requests to the upstream.
+        # Therefore, we may receive from the upstream gzip-encoded
+        # response along with "Content-Encoding:gzip" header.
+        # We either need to encode the response, or
+        # remove the "Content-Encoding" header.
+        if "Content-Encoding" in httpx_headers:
+            del httpx_headers["Content-Encoding"]
+
+        headers = {k.decode(): v.decode() for k, v in httpx_headers.raw}
+
+        return parse_adapter_exception(
             status_code=r.status_code,
             headers=headers,
+            content=r.text,
         )
 
-    if isinstance(e, APITimeoutError):
-        raise DialException(
+    if isinstance(exc, APITimeoutError):
+        return DialException(
             status_code=504,
             type="timeout",
             message="Request timed out",
             display_message="Request timed out. Please try again later.",
         )
 
-    if isinstance(e, APIConnectionError):
-        raise DialException(
+    if isinstance(exc, APIConnectionError):
+        return DialException(
             status_code=502,
             type="connection",
             message="Error communicating with OpenAI",
             display_message="OpenAI server is not responsive. Please try again later.",
         )
 
-    if isinstance(e, APIError):
-        raise DialException(
-            status_code=getattr(e, "status_code", None) or 500,
-            message=e.message,
-            type=e.type,
-            code=e.code,
-            param=e.param,
-            display_message=None,
+    if isinstance(exc, APIError):
+        # Streaming errors reported by `openai` library via this exception
+        status_code: int = 500
+        if exc.code:
+            try:
+                status_code = int(exc.code)
+            except Exception:
+                pass
+
+        return parse_adapter_exception(
+            status_code=status_code,
+            headers={},
+            content={"error": exc.body or {}},
         )
 
+    if isinstance(exc, DialException):
+        return exc
 
-def pydantic_exception_handler(request: Request, exc: pydantic.ValidationError):
-    return pydantic_validation_exception_handler(request, exc)
+    return InternalServerError(str(exc))
 
 
-def dial_exception_handler(request: Request, exc: DialException):
-    return exc.to_fastapi_response()
+def openai_exception_handler(request: Request, exc: Exception) -> Response:
```

Comment:
Maybe the same approach, as I proposed for DialException:

Replace this function with `assert` inside with just `lambda` right in exception handler:
```python
app.add_exception_handler(
   OpenAIError,
   lambda _, err: to_adapter_exception(err).to_fastapi_response()
)
```
What do you think?

Reply:
fixed